File: 2020/day/14/main.py
import fileinput

N = 36

class Mem1:
    def __init__(self, and_mask = 0, or_mask = 0):
        self.mem = {}
        self.and_mask = and_mask
        self.or_mask = or_mask

    def set_mask(self, mask):
        self.and_mask = 0
        self.or_mask = 0

        for i, x in enumerate(mask):
            if x == '0':
                pass
            elif x == '1':
                self.or_mask |= 1 << (N - i - 1)
            elif x == 'X':
                self.and_mask |= 1 << (N - i - 1)

    def sum_values(self):
        return sum(self.mem.values())

    def __setitem__(self, key, value):
        self.mem[key] = (value & self.and_mask) | self.or_mask


class Mem2:
    def __init__(self, mask = None):
        self.mem = {}
        self.mask = mask

    def sum_values(self):
        return sum(self.mem.values())

    def set_mask(self, mask):
        self.mask = mask

    def apply_mask(self, value):
        values = set([value])

        for i in range(N):
            new_values = set()
            x = self.mask[i]

            for v in values:
                if x == '0':
                    new_values.add(v)
                elif x == '1':
                    new_values.add(v | (1 << (N - i - 1)))
                elif x == 'X':
                    new_values.add(v & ~(1 << (N - i - 1)))
                    new_values.add(v | (1 << (N - i - 1)))

            values = new_values

        return values

    def __setitem__(self, key, value):
        for k in self.apply_mask(key):
            self.mem[k] = value


with fileinput.input('input.txt') as f:
    mem1 = Mem1()
    mem2 = Mem2()

    for line in f:
        if line.startswith('mask = '):
            mask = line[len('mask = '):-1]
            mem1.set_mask(mask)
            mem2.set_mask(mask)
        else:
            exec(line, None, {'mem': mem1})
            exec(line, None, {'mem': mem2})

# Part 1
print(mem1.sum_values())

# Part 2
print(mem2.sum_values())

File: 2020/day/14/Main.hs
import Text.ParserCombinators.ReadP
import Data.Char
import Data.Bits
import qualified Data.IntMap as IntMap

data Stmt
    = Mask String  -- mask = value
    | Mem Int Int  -- mem[key] = value
  deriving (Eq, Show)

numBits = 36

stmt = mask +++ mem

parse = stmt `endBy` char '\n' <* eof

mask :: ReadP Stmt
mask = Mask <$> (string "mask = " *> count numBits (satisfy (`elem` "01X")))

mem :: ReadP Stmt
mem = Mem <$> (string "mem" *> between (char '[') (char ']') number) <*> (string " = " *> number)

number = read <$> munch1 isDigit

eval1 (_, mem) (Mask newMask) = (Just newMask, mem)
eval1 (Just mask, mem) (Mem key value) = (Just mask, IntMap.insert key (applyMask value mask) mem)

applyMask :: Int -> String -> Int
applyMask value = snd . foldr f (0, 0)
  where
    f '0' (i, v) = (i + 1, v `clearBit` i)
    f '1' (i, v) = (i + 1, v `setBit` i)
    f 'X' (i, v)
        | value `testBit` i = (i + 1, v `setBit` i)
        | otherwise         = (i + 1, v `clearBit` i)

eval2 (_, mem) (Mask newMask)  = (Just newMask, mem)
eval2 (Just mask, mem) (Mem key value) =
    (Just mask, mem <> IntMap.fromList [(key', value) | key' <- expandMask value mask])

expandMask :: Int -> String -> [Int]
expandMask value = snd . foldr f (0, [value])
  where
    f '0' (i, vs) = (i + 1, vs)
    f '1' (i, vs) = (i + 1, map (`setBit` i) vs)
    f 'X' (i, vs) = (i + 1, map (`clearBit` i) vs <> map (`setBit` i) vs)

run step = sum . snd . foldl step (Nothing, mempty)

main = do
    input <- readFile "input.txt"
    let [(prog, "")] = readP_to_S parse input

    -- Part 1
    print $ run eval1 prog
    
    -- Part 2
    print $ run eval2 prog
